File: generate/testdata/dart/run_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

// These should be copied into place when we run 'make generate' for testing.
import 'sample_service.gen.client.dart';

main(List<String> args) async {
  if (args.length < 2) {
    print('Usage:');
    print('  dart run_client.dart {ADDRESS} {TEST_CASE}');
    print('');
    print('Example:');
    print('  dart run_client.dart localhost:9000 DownloadResumable');
    print('  dart run_client.dart localhost:9001 Authorization');
    exit(1);
  }
  await runTestCase(args[0], args[1]);
  exit(0);
}

runTestCase(String hostPort, String name) async {
  var baseURI = 'http://' + hostPort;

  switch (name) {
    case 'NotConnected':
      var client = new SampleServiceClient(baseURI);
      return output(client.Defaults(SampleRequest(text: 'Dude')));

    case 'Defaults':
      var client = new SampleServiceClient(baseURI);
      return output(client.Defaults(SampleRequest(text: 'Dude')));

    case 'ComplexValues':
      var client = new SampleServiceClient(baseURI);
      return output(client.ComplexValues(SampleComplexRequest(
        inFlag: true,
        inFloat: 3.14,
        inTime: '2022-12-05T17:47:12+00:00',
        inTimePtr: '2020-11-06T17:47:12+00:00',
        inUser: SampleUser(
          id: '123',
          name: 'Dude',
          age: 47,
          attention: 1000000,
          attentionString: '4m2s',
          digits: '555-1234',
          marshalToString: 'home@string.com,work@string.com',
          marshalToObject: {
            'H': 'home@object.com',
            'W': 'work@object.com',
          },
        ),
      )));

    case 'ComplexValuesPath':
      var client = new SampleServiceClient(baseURI);
      return output(client.ComplexValuesPath(SampleComplexRequest(
        inFlag: true,
        inFloat: 3.14,
        inTime: '2022-12-05T17:47:12+00:00',
        inTimePtr: '2020-11-06T17:47:12+00:00',
        inUser: SampleUser(
          id: '123',
          name: 'Dude',
          age: 47,
          attention: 1000000,
          attentionString: '4m2s',
          digits: '555-1234',
          marshalToString: 'home@string.com,work@string.com',
          marshalToObject: {
            'Home': "home@object.com",
            'Work': "work@object.com",
          },
        ),
      )));

    case 'Fail4XX':
      var client = new SampleServiceClient(baseURI);
      return output(client.Fail4XX(SampleRequest()));

    case 'Fail5XX':
      var client = new SampleServiceClient(baseURI);
      return output(client.Fail5XX(SampleRequest()));

    case 'CustomRoute':
      var client = new SampleServiceClient(baseURI);
      return output(client.CustomRoute(SampleRequest(id: '123', text: 'Abide')));

    case 'CustomRouteBody':
      var client = new SampleServiceClient(baseURI);
      return output(client.CustomRouteBody(SampleRequest(id: '123', text: 'Abide')));

    case 'CustomRouteQuery':
      var client = new SampleServiceClient(baseURI);
      return output(client.CustomRouteQuery(SampleRequest(id: '123', text: 'Abide')));

    case 'OmitMe':
      var client = new SampleServiceClient(baseURI);
      return output(client.OmitMe(SampleRequest()));

    case 'Download':
      var client = new SampleServiceClient(baseURI);
      return outputRaw(client.Download(SampleDownloadRequest(format: 'text/plain')));

    case 'DownloadResumable':
      var client = new SampleServiceClient(baseURI);
      return outputRaw(client.DownloadResumable(SampleDownloadRequest(format: 'text/plain')));

    case 'Redirect':
      var client = new SampleServiceClient(baseURI);
      return outputRaw(client.Redirect(SampleRedirectRequest()));

    case 'Authorization':
      var client = new SampleServiceClient(baseURI);
      return output(client.Authorization(SampleRequest(), authorization: 'Abide'));

    case 'AuthorizationGlobal':
      var client = new SampleServiceClient(baseURI, authorization: '12345');
      return output(client.Authorization(SampleRequest()));

    case 'AuthorizationOverride':
      var client = new SampleServiceClient(baseURI, authorization: '12345');
      return output(client.Authorization(SampleRequest(), authorization: 'Abide'));

    default:
      print('Unknown test case: "$name"');
      exit(1);
  }
}

output(Future<ModelJSON> model) async {
  try {
    var jsonString = jsonEncode(await model);
    print('OK ${jsonString}');
  }
  on SampleServiceException catch (err) {
    var message = err.message.replaceAll('"', '\'').trim();
    print('FAIL {"status":${err.status}, "message": "${message}"}');
  }
  catch (err) {
    print('FAIL {"message": "$err"}');
  }
}

outputRaw(Future<ModelStream> modelFuture) async {
  try {
    var model = (await modelFuture);
    var modelJson = model.toJson();

    // Content is still a future, so resolve it in order to print the raw data.
    modelJson['Content'] = await modelJson['Content'];
    print('OK ${jsonEncode(modelJson)}');
  }
  on SampleServiceException catch (err) {
    var message = err.message.replaceAll('"', '\'');
    print('FAIL {"status":${err.status}, "message": "${message}"}');
  }
  catch (err) {
    print('FAIL {"message": "$err"}');
  }
}

File: generate/testdata/js/run_client.js
/* global require, process */

// We don't want node to include output regarding fetch being experimental... just use it.
process.removeAllListeners('warning')

// These should be copied into place when we run 'make generate' for testing.
// const {SampleServiceClient} = require('./sample_service.gen.client');
import {SampleServiceClient} from './sample_service.gen.client.js';

async function main() {
    if (process.argv.length < 4) {
        console.info("Usage:");
        console.info("  node run_client.js {ADDRESS} {TEST_CASE}");
        console.info("");
        console.info("Example:");
        console.info("  node run_client.js localhost:9000 DownloadResumable");
        console.info("  node run_client.js localhost:9001 Authorization");
    }

    const baseURI = 'http://' + process.argv[2];
    const testCase = process.argv[3];

    switch (testCase) {
        case 'NotConnected': {
            const client = new SampleServiceClient(baseURI);
            return output(client.Defaults({Text: 'Hello'}));
        }
        case 'BadFetch': {
            const client = new SampleServiceClient(baseURI, {fetch: 'fart'});
            return output(client.Defaults({Text: 'Hello'}));
        }
        case 'Defaults': {
            const client = new SampleServiceClient(baseURI);
            return output(client.Defaults({Text: 'Hello'}));
        }
        case 'ComplexValues': {
            const client = new SampleServiceClient(baseURI);
            return output(client.ComplexValues({
                InUser: {
                    ID: '123',
                    Name: 'Dude',
                    Age: 47,
                    Attention: 1000000,
                    AttentionString: '4m2s',
                    Digits: '555-1234',
                    MarshalToString: "home@string.com,work@string.com",
                    MarshalToObject: { H: "home@object.com", W: "work@object.com" },
                },
                InFlag: true,
                InFloat: 3.14,
                InTime: '2022-12-05T17:47:12+00:00',
                InTimePtr: '2020-11-06T17:47:12+00:00',
            }));
        }
        case 'ComplexValuesPath': {
            const client = new SampleServiceClient(baseURI);
            return output(client.ComplexValuesPath({
                InUser: {
                    ID: '123',
                    Name: 'Dude',
                    Age: 47,
                    Attention: 1000000,
                    AttentionString: '4m2s',
                    Digits: '555-1234',
                    MarshalToString: "home@string.com,work@string.com",
                    MarshalToObject: { H: "home@object.com", W: "work@object.com" },
                },
                InFlag: true,
                InFloat: 3.14,
                InTime: '2022-12-05T17:47:12+00:00',
                InTimePtr: '2020-11-06T17:47:12+00:00',
            }));
        }
        case 'Fail4XX': {
            const client = new SampleServiceClient(baseURI);
            return output(client.Fail4XX({}));
        }
        case 'Fail5XX': {
            const client = new SampleServiceClient(baseURI);
            return output(client.Fail5XX({}));
        }
        case 'CustomRoute': {
            const client = new SampleServiceClient(baseURI);
            return output(client.CustomRoute({ID: '123', Text: 'Abide'}));
        }
        case 'CustomRouteQuery': {
            const client = new SampleServiceClient(baseURI);
            return output(client.CustomRouteQuery({ID: '456', Text: 'Abide'}));
        }
        case 'CustomRouteBody': {
            const client = new SampleServiceClient(baseURI);
            return output(client.CustomRouteBody({ID: '789', Text: 'Abide'}));
        }
        case 'OmitMe': {
            const client = new SampleServiceClient(baseURI);
            return output(client.OmitMe({}));
        }
        case 'Download': {
            const client = new SampleServiceClient(baseURI);
            return output(client.Download({Format: 'text/plain'}));
        }
        case 'DownloadResumable': {
            const client = new SampleServiceClient(baseURI);
            return output(client.DownloadResumable({Format: 'text/plain'}));
        }
        case 'Redirect': {
            const client = new SampleServiceClient(baseURI);
            return output(client.Redirect({}));
        }
        case 'Authorization': {
            const client = new SampleServiceClient(baseURI);
            return output(client.Authorization({}, {authorization: 'Abide'}));
        }
        case 'AuthorizationGlobal': {
            const client = new SampleServiceClient(baseURI, {authorization: '12345'});
            return output(client.Authorization({}));
        }
        case 'AuthorizationOverride': {
            const client = new SampleServiceClient(baseURI, {authorization: '12345'});
            return output(client.Authorization({}, {authorization: 'Abide'}));
        }
    }
}

async function output(responseFuture) {
    try {
        const value = await responseFuture;
        const content = value && value['Content'];

        // Duck typing to determine if we received a raw Blob back instead of decoded JSON. Our
        // raw stream responses all return byte streams of text, so just flatten to that value.
        if (content && content['arrayBuffer'] && content['text']) {
            value.Content = await value.Content.text();
        }
        console.info('OK ' + JSON.stringify(await value));
    }
    catch (e) {
        const failure = await e;
        const failureJSON = typeof failure === 'string'
            ? JSON.stringify({message: failure})
            : JSON.stringify(failure);

        console.info('FAIL ' + failureJSON);
    }
}

main().then().catch((e) => console.info('FAILURE:' + e));
